aten/src/ATen/native/cpu/batch_norm_kernel.cpp

```diff
@@ -87,28 +105,596 @@ void batch_norm_cpu_inference_contiguous_impl(Tensor& output,
+          Vec output_vec = data_vec * alpha_vec + beta_vec;
+          output_vec.store(output_data + offset + d, n_channel - d);
+        }
       }
-    }
+    });
   }
 }
 
-void batch_norm_cpu_inference_contiguous_kernel(Tensor& output, const Tensor& input,
-    const Tensor& weight, const Tensor& bias, const Tensor& mean, const Tensor& variance, double eps) {
-  AT_DISPATCH_FLOATING_TYPES(input.scalar_type(), "batch_norm_cpu_inference_contiguous", [&] {
-    batch_norm_cpu_inference_contiguous_impl<scalar_t>(output, input, weight, bias, mean, variance, eps);
+template <typename scalar_t>
+void batch_norm_cpu_channels_last_impl(Tensor& output, const Tensor& input,
+    const Tensor& weight, const Tensor& bias, const Tensor& save_mean, const Tensor& save_invstd,
+    const Tensor& running_mean, const Tensor& runnning_var, bool train, double eps) {
+
+  using Vec = Vec256<scalar_t>;
+  int64_t n_batch = input.size(0);
+  int64_t n_channel = input.size(1);
+  int64_t image_size = input.numel() / n_batch / n_channel;
+
+  Tensor alpha = at::empty({n_channel}, input.options());
+  Tensor beta = at::empty({n_channel}, input.options());
+  scalar_t* alpha_data = alpha.data_ptr<scalar_t>();
+  scalar_t* beta_data = beta.data_ptr<scalar_t>();
+
+  batch_norm_cpu_collect_linear_and_constant_terms<scalar_t>(
+      alpha_data, beta_data, n_channel, weight, bias,
+      save_mean, save_invstd, running_mean, runnning_var, train, eps);
+
+  scalar_t* output_data = output.data_ptr<scalar_t>();
+  const scalar_t* input_data = input.data_ptr<scalar_t>();
+
+  // Apply the linear terms to the input,
+  // output(n, c, h, w) = input(n, c, h, w) * alpha(c) + beta(c)
+  const int64_t loop_size = n_channel - (n_channel % Vec::size());
+  at::parallel_for(0, n_batch * image_size, 1, [&](int64_t begin, int64_t end) {
+    for (int64_t i = begin; i < end; i++) {
+      int64_t offset = i * n_channel;
+      int64_t d = 0;
+      // vectorize on channel dimension, for normal batch_norm input size,
+      // alpha/beta should fit in L1 cache, otherwise consider blocking.
+      for (; d < loop_size; d += Vec::size()) {
+        Vec alpha_vec = Vec::loadu(alpha_data + d);
+        Vec beta_vec = Vec::loadu(beta_data + d);
+        Vec data_vec = Vec::loadu(input_data + offset + d);
+        Vec output_vec = data_vec * alpha_vec + beta_vec;
+        output_vec.store(output_data + offset + d);
+      }
+      if (n_channel - d > 0) {
+        Vec alpha_vec = Vec::loadu(alpha_data + d, n_channel - d);
+        Vec beta_vec = Vec::loadu(beta_data + d, n_channel - d);
+        Vec data_vec = Vec::loadu(input_data + offset + d, n_channel - d);
+        Vec output_vec = data_vec * alpha_vec + beta_vec;
+        output_vec.store(output_data + offset + d, n_channel - d);
+      }
+    }
+  });
+}
+
+template <typename scalar_t>
+void batch_norm_cpu_collect_stats_contiguous_impl(
+    Tensor& mean, Tensor& var_sum, const Tensor& input) {
+
+  using Vec = Vec256<scalar_t>;
```

Comment:
Please remove this line. This is not used by any code within the function and caused a internal test failure.

Reply:
Fixed! Please review!